### ISSUE TITLE
Restore twice will let the spikes in queue vanish.

### ISSUE THREAD

**Issue by zhouyanasd** [brian-team/brian2#938]
Hello, I just have a confuse when I use Brian2, if I set up an SNN and then I run for some ms. Next, I store it using net.store('first_run'), but when I call the net.restore(‘first_run’) twice, I can not get the same response of neuron compare with I directly run next 1ms. 

I have checked the source code of the Synapticpathway class, the function of _restore_from_full_state() is like this: 
```    
def _restore_from_full_state(self, state):
            queue_state = state.pop('_spikequeue', None)
            super(SynapticPathway, self)._restore_from_full_state(state)
            if self.queue is None:
                 self.queue = get_device().spike_queue(self.source.start, self.source.stop)
            self.queue._restore_from_full_state(queue_state)
```
I think the state.pop is the main problem of what I concer, would you please tell me is it a mistake or just   write it on purpose?

Thank you very much !!  
And, I want to join the developers If you still need some people in this project.

**Comment by mstimberg** [comment]
I agree that the `pop` looks wrong there... Could you create a minimal example that shows this issue?

> And, I want to join the developers If you still need some people in this project.

There's no formal process to join the development team, but contributions are always welcome! E.g. if you can debug this error and fix it, please open a pull request.

**Comment by zhouyanasd** [comment]
Sure, a demo of illustrating this bug is as fellows:
```python
from brian2 import *

start_scope()
prefs.codegen.target = "numpy"
defaultclock.dt = 1 * ms

equ = '''
dv/dt = (I-v) / (30*ms) : 1 (unless refractory)
dh/dt = (-h)/(6*ms) : 1
I = (h)+13.5: 1
'''

on_pre = '''
h+=w
'''
stimulus = TimedArray(np.tile([20., 0.], 5)*Hz, dt=1000.*ms)
P = PoissonGroup(1, rates='stimulus(t)')

G = NeuronGroup(10, equ, threshold='v > 15', reset='v = 13.5', method='euler', refractory=3 * ms,
                name='neurongroup_ex')
S_in = Synapses(P, G, 'w : 1', on_pre=on_pre, method='euler', name='synapses_in', delay=1.5 * ms)
S_e = Synapses(G, G, 'w : 1', on_pre=on_pre, method='euler', name='synapses_E')
S_in.connect(p=0.3)
S_e.connect(condition='i!=j', p=0.5)

G.v = '13.5+1.5*rand()'
G.h = '0'
S_in.w = 'randn()*30+30'
S_e.w = 'randn()*30+30'
S_e.delay='randn()*3 * ms+3*ms'

G.run_regularly(''' lastspike = 0 * ms
                     v = 13.5+1.5*rand()
                    h = 0
                     ''', dt=1000 * ms)

net = Network(collect())
net.store('init')
net.run(1000*ms, report='text')
net.store('temp')
```
and then I will run:
```python
net.run(1*ms)
print(G.v)
```
I got:
```
<neurongroup_ex.h: array([   0.        ,    0.        ,   56.73154665,  -15.51187815,
          0.        ,  -32.40821256,    0.        ,   61.18894273,
        115.25546651,   63.88970469])>
```
After that, I run the fellow codes once:
```python
net.restore('temp')
net.run(1*ms)
G.h
```
 I got the same results.
But when I run `net.restore('temp')`  the second time , I just got the different results:
```
<neurongroup_ex.h: array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.])>
```
I have attached my jupyter notebook file, you can run and test it.

**Comment by mstimberg** [comment]
Thanks for the example, I can confirm the issue and I think your analysis is correct. The problem is that spikes that are still "in the queue" (i.e. not yet delivered because of the synaptic delay) are removed by the `restore` function. Here's a shorter example showing the same problem:
```Python
from brian2 import *

source = SpikeGeneratorGroup(1, [0], [0]*ms)
target = NeuronGroup(1, 'v : 1')
conn = Synapses(source, target, on_pre='v += 1', delay=2*defaultclock.dt)
conn.connect()
run(defaultclock.dt)
store()
run(2*defaultclock.dt)
print(target.v)
restore()
run(2*defaultclock.dt)
print(target.v)
restore()
run(2*defaultclock.dt)
print(target.v)
```

Happy to let you fix it and open a pull request :smile:

**Comment by zhouyanasd** [comment]
OK, will do. Thank you for the reply and I will pull request as soon as possible!^_^

**Comment by mstimberg** [comment]
There was actually a reason for removing the spike queue from the state dictionary (otherwise it would be treated like any state variable, and the standard restore process in `VariableOwner` would fail), the fix is to put it back again afterwards. I submitted a fix with PR #960.
